token-lending/program/src/processor.rs

```diff
@@ -1343,130 +1323,76 @@ fn process_withdraw_obligation_collateral(
-        return Err(LendingError::ReserveCollateralDisabled.into());
+    if repay_reserve.last_update.is_stale(clock.slot)? {
+        return Err(LendingError::ReserveStale.into());
     }
 
-    if withdraw_reserve_info.key == borrow_reserve_info.key {
-        return Err(LendingError::DuplicateReserve.into());
+    let withdraw_reserve = Reserve::unpack(&withdraw_reserve_info.data.borrow())?;
+    if withdraw_reserve_info.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
     }
-
-    if &withdraw_reserve.collateral.supply_pubkey != source_collateral_info.key {
-        msg!("Invalid withdraw reserve collateral supply account input");
+    if &withdraw_reserve.lending_market != lending_market_info.key {
+        msg!("Invalid withdraw reserve lending market account");
         return Err(LendingError::InvalidAccountInput.into());
     }
-
-    if &withdraw_reserve.collateral.supply_pubkey == destination_collateral_info.key {
-        msg!(
-            "Cannot use withdraw reserve collateral supply as destination collateral account input"
-        );
+    if &withdraw_reserve.collateral.supply_pubkey != withdraw_reserve_collateral_supply_info.key {
+        msg!("Invalid withdraw reserve collateral supply account");
         return Err(LendingError::InvalidAccountInput.into());
     }
-
-    // TODO: handle case when neither reserve is the quote currency
-    if borrow_reserve.dex_market.is_none() && withdraw_reserve.dex_market.is_none() {
-        msg!("One reserve must have a dex market");
+    if &withdraw_reserve.collateral.supply_pubkey == destination_collateral_info.key {
+        msg!("Invalid destination collateral account");
         return Err(LendingError::InvalidAccountInput.into());
     }
-    if let COption::Some(dex_market_pubkey) = borrow_reserve.dex_market {
-        if &dex_market_pubkey != dex_market_info.key {
-            msg!("Invalid dex market account input");
-            return Err(LendingError::InvalidAccountInput.into());
-        }
+    if withdraw_reserve.last_update.is_stale(clock.slot)? {
+        return Err(LendingError::ReserveStale.into());
     }
-    if let COption::Some(dex_market_pubkey) = withdraw_reserve.dex_market {
-        if &dex_market_pubkey != dex_market_info.key {
-            msg!("Invalid dex market account input");
-            return Err(LendingError::InvalidAccountInput.into());
-        }
+
+    // @TODO: what if a user borrows using the same reserve & mint for collateral?
+    //        this is permitted in borrow_obligation_liquidity and could be used for leverage.
+    //        maybe liquidation should be allowed, but reduce/eliminate the bonus.
+    if repay_reserve_info.key == withdraw_reserve_info.key {
+        return Err(LendingError::DuplicateReserve.into());
+    }
+    if repay_reserve.liquidity.mint_pubkey == withdraw_reserve.liquidity.mint_pubkey {
+        return Err(LendingError::DuplicateReserveMint.into());
     }
 
```

Comment:
@jstarry Can we discuss the desired behavior of this?

Reply:
@jordansexton Are you specifically asking about whether a liquidator can liquidate an obligation's loan in return for collateral of the same currency? I believe this should be allowed since it could happen for a purely leveraged case where a loan's interest causes the obligation to become unhealthy, right? I think the bonus is still needed because there needs to be an incentive for liquidators to take action on unhealthy loans. It probably doesn't make sense for the bonus to be as high as it would be for a more volatile pair, but I think that can be solved later if needed.

Reply:
Okay, cool. The current behavior is that a liquidation cannot repay and withdraw with the same reserve and mint. I just waned to check if this should be allowed (I think it should). I'll remove these checks.

Reply:
Okay, cool. The current behavior is that a liquidation cannot repay and withdraw with the same reserve and mint. I just wanted to check if this should be allowed (I think it should). I'll remove these checks.